### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by slendersold**

### CODE REVIEW

File: Untitled-1.ipynb
Comment (aaptss) [inline Untitled-1.ipynb:1]: что за название такое

File: src/colour_lib/palette_parser/ImageAlignment.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+            self.position["flip_horizontal"] = True
+            flipped = cv2.flip(image_to_flip, 1)
+        elif not y1000_more_than_yCA and not x1000_more_than_x500:
+            self.position["flip_vertical"] = True
+            flipped = cv2.flip(image_to_flip, 0)
+        elif not y1000_more_than_yCA and x1000_more_than_x500:
+            self.position["flip_over"] = True
+            flipped = cv2.rotate(image_to_flip, cv2.ROTATE_180)
+        else:
+            flipped = image_to_flip
```
Comment (aaptss) [inline src/colour_lib/palette_parser/ImageAlignment.py:34]: это идёт в самое начало и тогла этот else не нужен. при этом в других if-ах `flipped = operation(flipped)`

File: src/colour_lib/palette_parser/ImageAlignment.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+            if (
+                np.pi / 2 - 0.1 < theta < np.pi / 2 + 0.1
+                or 3 * np.pi / 2 - 0.1 < theta < 3 * np.pi / 2 + 0.1
+            ):
+                theta_array.append(theta)
+        theta_deg_mean = np.mean(np.rad2deg(theta_array))
+        angle_to_rotate = 90 - theta_deg_mean
+        self.position["rotation_angle"] = angle_to_rotate
+
+        center = (w / 2, h / 2)
```
Comment (aaptss) [inline src/colour_lib/palette_parser/ImageAlignment.py:57]: здесь идет работа с координатами? если w или h нечетные то выйдет нецелое число. consider using floor division `//`

File: src/colour_lib/palette_parser/ImageAlignment.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+        gray = cv2.cvtColor(image_to_rotate, cv2.COLOR_BGR2GRAY)
+        edges = cv2.Canny(gray, 50, 150, apertureSize=3)
+        lines = cv2.HoughLines(edges, 1, np.pi / 180, 200)
+
+        theta_array = []
+        for r_theta in lines:
+            r, theta = np.array(r_theta[0], dtype=np.float64)
+            # if normal is vertical (means that detected lines are horizontal)
+            if (
+                np.pi / 2 - 0.1 < theta < np.pi / 2 + 0.1
```
Comment (aaptss) [inline src/colour_lib/palette_parser/ImageAlignment.py:49]: неясно что за константа 0.1. нужен или комментарий или она должна быть переменной в окружении с понятным названием

File: src/colour_lib/palette_parser/ImageAlignment.py [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+                # Step 2: Apply rotation matrix
+                x_rotated = x_origin * np.cos(theta) - y_origin * np.sin(theta)
+                y_rotated = x_origin * np.sin(theta) + y_origin * np.cos(theta)
+                # Step 3: Translate point back
+                x_new = x_rotated + cx
+                y_new = y_rotated + cy
+
+                return x_new, y_new
+
+            cx = int(width / 2)
```
Comment (aaptss) [inline src/colour_lib/palette_parser/ImageAlignment.py:153]: width // 2

File: src/colour_lib/palette_parser/ImageAlignment.py [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+        return image_rotated, image_to_show
+
+    def give_original_coordinates(self, coordinates, w, h, type):
+
+        def flip_coordinates_horizontal(coordinates, width, type):
+
+            flipped_dict = {}
+
+            if type == "circle":
+                for key, value in coordinates.items():
```
Comment (aaptss) [inline src/colour_lib/palette_parser/ImageAlignment.py:90]: итератор можно вынести вовне иф стейтмента потому что он одинаков

File: src/colour_lib/palette_parser/ImageProcessing.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+        Parameters:
+        image (numpy array): The input image.
+
+        Returns:
+        numpy array: The image with enhanced contrast.
+        """
+        lab = cv2.cvtColor(image, cv2.COLOR_BGR2LAB)
+        l, a, b = cv2.split(lab)
+        clahe = cv2.createCLAHE(clipLimit=3.0, tileGridSize=(8, 8))
```
Comment (aaptss) [inline src/colour_lib/palette_parser/ImageProcessing.py:52]: tileGridSize по хорошему передавать в качестве аргумента в метод с указанием значения по умолчанию, то есть 

```
def clahe(self, image, tileGridSize=(8, 8)):
    ...
    clahe = cv2.createCLAHE(clipLimit=3.0, tileGridSize=tileGridSize)
```

File: src/colour_lib/palette_parser/ImageProcessing.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+        Parameters:
+        image (numpy array): The input image.
+
+        Returns:
+        numpy array: The edges detected in the image.
+        """
+
+        gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
+        blurred = cv2.GaussianBlur(gray, (9, 9), 0)
```
Comment (aaptss) [inline src/colour_lib/palette_parser/ImageProcessing.py]: аналогично, такие вещи как размеры ядер должны быть параметрами которые можно эксплицитно передать извне

File: src/colour_lib/palette_parser/ObjectColor.py [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+
+class ObjectColor:
+    def __init__(self):
+        pass
+
+    def calculate_rgb(self, coordinates, zarr, filter="_1000"):
+
+        def mean_rgb_circ(image, x, y, r, margin=0.3):
+
+            r = int(r * margin)
```
Comment (aaptss) [inline src/colour_lib/palette_parser/ObjectColor.py:13]: margin=0.3 и тайпкаст в инты выглядит как рецепт бага

File: src/colour_lib/palette_parser/ObjectColor.py [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+
+            Y, X = np.ogrid[: image.shape[0], : image.shape[1]]
+
+            dist_from_center = np.sqrt((X - x) ** 2 + (Y - y) ** 2)
```
Comment (aaptss) [inline src/colour_lib/palette_parser/ObjectColor.py:17]: что такое большой икс и малый икс, как это понимать

File: src/colour_lib/palette_parser/ObjectColor.py [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+        # Display the grid
+        ax.imshow(color_grid, aspect="auto")
+
+        # Set the tick labels
+        ax.set_xticks(np.arange(4))
+        ax.set_xticklabels(["A", "B", "C", "D"])
+        ax.set_yticks(np.arange(7))
+        ax.set_yticklabels(["CA", "6", "5", "4", "3", "2", "1"])
+
+        plt.show()
```
Comment (aaptss) [inline src/colour_lib/palette_parser/ObjectColor.py]: не надо так делать. 

`return ax`

File: src/colour_lib/palette_parser/ObjectColor.py [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+        coordinates_filtered = {
+            key: value
+            for key, value in coordinates.items()
+            if (filter in key) and ("rect" not in key)
+        }
+        coordinates_filtered["rect_CA"] = coordinates["rect_CA"]
+
+        obj_rgb = {}
+        for i, (key, values) in enumerate(coordinates_filtered.items()):
+            print(key)
```
Comment (aaptss) [inline src/colour_lib/palette_parser/ObjectColor.py:54]: дебаг-принты следует убрать. 
не только здесь, но и вообще в целом. 

если это не дебаг-принт, то вывод должен быть осмысленным, как в ImageAlignment реализовано
image

File: src/colour_lib/palette_parser/ObjectColor.py [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+
+        Args:
+        obj_rgb (dict): Dictionary containing mean RGB values for each object.
+        """
+        fig, ax = plt.subplots(figsize=(4, 7))
+
+        # Create a grid of colors based on the obj_rgb keys
+        color_grid = np.zeros(
+            (7, 4, 3)
+        )  # Assuming 7 rows and 4 columns for CA and A-D with 1-6
```
Comment (aaptss) [inline src/colour_lib/palette_parser/ObjectColor.py:94]: асьюмить не надо, надо передать аргументом с дефолтным параметром

Comment (aaptss) [reply src/colour_lib/palette_parser/ObjectColor.py:94]: тогда и хардкоды всяких `range(4)` и `row_index = 7 - ...` пропадут

File: src/colour_lib/palette_parser/ObjectColor.py [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+
+        # Normalize the color values to [0, 1] for displaying with imshow
+        color_grid /= 255.0
+
+        # Display the grid
+        ax.imshow(color_grid, aspect="auto")
+
+        # Set the tick labels
+        ax.set_xticks(np.arange(4))
+        ax.set_xticklabels(["A", "B", "C", "D"])
```
Comment (aaptss) [inline src/colour_lib/palette_parser/ObjectColor.py:123]: тоже есть смысл передать как арг функции

File: src/colour_lib/palette_parser/ObjectDetection.py [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+        ]
+
+        # Combine areas and vertices for sorting
+        combined = list(zip(areas, vertices_new_format))
+
+        # Sort the combined list by area in descending order
+        combined.sort(key=lambda x: x[0], reverse=True)
+
+        # Define keys for the rectangles
+        keys = ["rect_CA", "rect_1000", "rect_750", "rect_500", "rect_400", "rect_300"]
```
Comment (aaptss) [inline src/colour_lib/palette_parser/ObjectDetection.py]: хардкод -- плохо. передать аргументом функции.

File: src/colour_lib/palette_parser/ObjectDetection.py [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+            if i < len(combined):
+                area, (y0, y1, x0, x1) = combined[i]
+                coordinates[key] = {"y0": y0, "y1": y1, "x0": x0, "x1": x1}
+
+        # 3. SHOW INPUT IMAGE WITH DETECTED RECTANGLES (OPTIONAL)
+
+        if show_image:
+            image_to_show_ = image_to_show.copy()
+            for rect in vertices:
+                cv2.drawContours(image_to_show_, [rect], -1, (250, 199, 16), 20)
```
Comment (aaptss) [inline src/colour_lib/palette_parser/ObjectDetection.py:75]: хардкод -- плохо. передать аргументом функции.

File: src/colour_lib/palette_parser/ObjectDetection.py [hunk 16]
```diff
@@ -0,0 +1,187 @@
...
+        else:
+            return coordinates, None
+
+    def find_circles(self, image, averaging_threshold):
+
+        def is_circle(contour, width, tolerance=0.2):
+
+            perimeter = cv2.arcLength(contour, True)
+            area = cv2.contourArea(contour)
+            if perimeter < int(width / 6):
```
Comment (aaptss) [inline src/colour_lib/palette_parser/ObjectDetection.py:86]: что за /6

File: src/colour_lib/palette_parser/ObjectDetection.py [hunk 17]
```diff
@@ -0,0 +1,187 @@
...
+                return False
+            circularity = 4 * np.pi * (area / (perimeter * perimeter))
```
Comment (aaptss) [inline src/colour_lib/palette_parser/ObjectDetection.py:88]: что за 4пи 

на такие штуки следует давать или комментарии или такие константы задавать как переменные с говорящими нащваниями

File: src/colour_lib/palette_parser/ObjectDetection.py [hunk 18]
```diff
@@ -0,0 +1,187 @@
...
+            return 1 - tolerance <= circularity <= 1 + tolerance
+
+        def average(coordinates, threshold=10):
```
Comment (aaptss) [inline src/colour_lib/palette_parser/ObjectDetection.py:91]:
```
def average(coordinates, threshold=10):
    coordinates = np.array(coordinates)
    num_coords = len(coordinates)
    
    distances = np.linalg.norm(coordinates[:, np.newaxis] - coordinates, axis=2)
    
    groups = -np.ones(num_coords, dtype=int)
    group_count = 0

    for i in range(num_coords):
        if groups[i] == -1:  # if this coordinate is not yet assigned to a group
            within_threshold = distances[i] <= threshold
            groups[within_threshold] = group_count
            group_count += 1
    
    unique_groups = [coordinates[groups == i] for i in range(group_count)]
    mean_values = [np.mean(group, axis=0).astype(int).tolist() for group in unique_groups]
    
    return mean_values
  ```

Comment (aaptss) [reply src/colour_lib/palette_parser/ObjectDetection.py:91]: векторизовано, и не нужно никаких поэлементных вычислений расстояний

File: src/colour_lib/palette_parser/ObjectDetection.py [hunk 19]
```diff
@@ -0,0 +1,187 @@
...
+
+        # Write circle coordinates
+        circles_coord = []
+        for contour in circles:
+            (x, y), radius = cv2.minEnclosingCircle(contour)
+            center = (int(x), int(y))
+            radius = int(radius)
+            circles_coord.append([center[0], center[1], radius])
+
+        mean_radius = int(np.mean([c[2] for c in circles_coord]))  #
```
Comment (aaptss) [inline src/colour_lib/palette_parser/ObjectDetection.py:131]: лишние # убрать

File: src/colour_lib/palette_parser/ObjectDetection.py [hunk 20]
```diff
@@ -0,0 +1,187 @@
...
+
+        # Sort each subgroup of 4 circles by x-coordinate
+        n = 4
+        for i in range(0, len(sorted_by_y), n):
+            subgroup = sorted_by_y[i : i + n]
+            sorted_subgroup = sorted(subgroup, key=lambda item: item[0])
+            sorted_by_y[i : i + n] = sorted_subgroup
+
+        # Characters and numbers for generating keys
+        characters = ["A", "B", "C", "D"]
```
Comment (aaptss) [inline src/colour_lib/palette_parser/ObjectDetection.py:151]: на этом уровне уже можно даже сказать не "передайте аргументами" а в либе просто выделите константу CHARACTERS = [ABCD] и импортируйте её где применимо

File: src/colour_lib/regressors/LassoRegressor.py [hunk 21]
```diff
@@ -0,0 +1,11 @@
+from sklearn import linear_model
+from ..regressors.AbstractRegressor import AbstractRegressor
```
Comment (aaptss) [inline src/colour_lib/regressors/LassoRegressor.py]: надо обойтись без относительных импортов

File: src/colour_lib/regressors/VoteRegressor.py [hunk 22]
```diff
@@ -0,0 +1,37 @@
...
+from sklearn.linear_model import LinearRegression
+from sklearn.ensemble import RandomForestRegressor
+
+# from sklearn.svm import SVR
+from sklearn.multioutput import MultiOutputRegressor
+
+
+class VoteRegressor(AbstractRegressor):
+    def __init__(self, train_data, reference_data, **kwargs):
+        # base_models = [
```
Comment (aaptss) [inline src/colour_lib/regressors/VoteRegressor.py]: неиспользуемые комментарии убрать

File: src/colour_lib/utils/rawparser.py [hunk 23]
```diff
@@ -0,0 +1,132 @@
...
+    MSDS_CMFS,
+    SDS_ILLUMINANTS,
+    MultiSpectralDistributions,
+    SpectralShape,
+    gamma_function,
+    msds_to_XYZ,
+)
+from colour.models.rgb import RGB_COLOURSPACES, XYZ_to_RGB
+
+default_colref_basepath = "."
```
Comment (aaptss) [inline src/colour_lib/utils/rawparser.py]: оно нигде не используется. зачем оно есть?

File: src/colour_lib/utils/rawparser.py [hunk 24]
```diff
@@ -0,0 +1,132 @@
...
+        return self.calculate_xyz_chart(
+            self.msds,
+            observer=MSDS_CMFS["CIE 1931 2 Degree Standard Observer"],
+            illuminant=SDS_ILLUMINANTS["D65"],
+        )
+
+    def get_reference_srgbs(self):
+        return self.calculate_rgb_chart(self.get_reference_d65(), "sRGB")
+
+    def get_reference_don4s(self):
```
Comment (aaptss) [inline src/colour_lib/utils/rawparser.py]: дон4 нигде не используется, можно и грохнуть

File: src/colour_lib/utils/rawparser.py [hunk 25]
```diff
@@ -0,0 +1,132 @@
...
+        )
+
+    def get_reference_d65(self):
+        return self.calculate_xyz_chart(
+            self.msds,
+            observer=MSDS_CMFS["CIE 1931 2 Degree Standard Observer"],
+            illuminant=SDS_ILLUMINANTS["D65"],
+        )
+
+    def get_reference_srgbs(self):
```
Comment (aaptss) [inline src/colour_lib/utils/rawparser.py]: не хватает дженерик метода который бы брал референс произвольного колорспейса, который уточняется аргументом передаваемым в функцию

File: CCexps.ipynb
Comment (aaptss) [inline CCexps.ipynb:1]: на весь ноутбук дам один коммент: для отображения картинок рисочки с координатами по осям Х и У не нужны, поэтому axis off

File: palette/parser_Cold_Calib_Polaris_Scan3.ipynb
Comment (aaptss) [inline palette/parser_Cold_Calib_Polaris_Scan3.ipynb:1]: а это обязательно должны быть ноутбуки? кажется у них одинаковый код, значит из этого нужно сделать скрипты с аргументами через argparse и запускать их для всех картьинок подряд. можно в ридми добавить коллы скриптов для каждого из изображений.

олсо, советую убрать понятия Polaris и Huron, чтоб сторонний читатель либы мог не чесал репу

**Review by aaptss — CHANGES_REQUESTED** [review]
